Mathlib/AlgebraicGeometry/Stalk.lean

```diff
@@ -51,12 +61,264 @@ theorem IsAffineOpen.fromSpecStalk_eq {X : Scheme} (x : X) {U V : X.Opens}
+  obtain ⟨_, ⟨U, hU, rfl⟩, hxU, -⟩ := (isBasis_affine_open Y).exists_subset_of_mem_open
+    (Set.mem_univ (f.1.base x)) isOpen_univ
+  obtain ⟨_, ⟨V, hV, rfl⟩, hxV, hVU⟩ := (isBasis_affine_open X).exists_subset_of_mem_open
+    hxU (f ⁻¹ᵁ U).2
+  rw [← hU.fromSpecStalk_eq_fromSpecStalk hxU, ← hV.fromSpecStalk_eq_fromSpecStalk hxV,
+    IsAffineOpen.fromSpecStalk, ← Spec.map_comp_assoc, Scheme.stalkMap_germ f _ x hxU,
+    IsAffineOpen.fromSpecStalk, Spec.map_comp_assoc, ← X.presheaf.germ_res (homOfLE hVU) x hxV,
+    Spec.map_comp_assoc, Category.assoc, ← Spec.map_comp_assoc (f.app _),
+      Hom.app_eq_appLE, Hom.appLE_map, IsAffineOpen.Spec_map_appLE_fromSpec]
+
+lemma Spec_fromSpecStalk (R : CommRingCat) (x) :
+    (Spec R).fromSpecStalk x =
+      Spec.map ((ΓSpecIso R).inv ≫ (Spec R).presheaf.germ ⊤ x trivial) := by
+  rw [← (isAffineOpen_top (Spec R)).fromSpecStalk_eq_fromSpecStalk (x := x) trivial,
+    IsAffineOpen.fromSpecStalk, IsAffineOpen.fromSpec_top, isoSpec_Spec_inv,
+    ← Spec.map_comp]
+
+-- This is not a simp lemma to respect the abstraction boundaries
+/-- A variant of `Spec_fromSpecStalk` that breaks abstraction boundaries. -/
+lemma Spec_fromSpecStalk' (R : CommRingCat) (x) :
+    (Spec R).fromSpecStalk x = Spec.map (StructureSheaf.toStalk R _) :=
+  Spec_fromSpecStalk _ _
+
+/-- https://stacks.math.columbia.edu/tag/01J7 -/
+lemma range_fromSpecStalk {x : X} :
+    Set.range (X.fromSpecStalk x).1.base = { y | y ⤳ x } := by
+  ext y
+  constructor
+  · rintro ⟨y, rfl⟩
+    exact ((LocalRing.specializes_closedPoint y).map (X.fromSpecStalk x).1.base.2).trans
+      (specializes_of_eq fromSpecStalk_closedPoint)
+  · rintro (hy : y ⤳ x)
+    have := fromSpecStalk_closedPoint (x := y)
+    rw [← stalkSpecializes_fromSpecStalk hy] at this
+    exact ⟨_, this⟩
+
+end Scheme
+
+end fromSpecStalk
+
+variable (R : CommRingCat.{u}) [LocalRing R]
+
+section stalkClosedPointIso
+
+/-- For a local ring `(R, 𝔪)`,
+this is the isomorphism between the stalk of `Spec R` at `𝔪` and `R`. -/
+noncomputable
+def stalkClosedPointIso :
+    (Spec R).presheaf.stalk (closedPoint R) ≅ R :=
+  StructureSheaf.stalkIso _ _ ≪≫ (IsLocalization.atUnits R
+      (closedPoint R).asIdeal.primeCompl fun _ ↦ not_not.mp).toRingEquiv.toCommRingCatIso.symm
+
+lemma stalkClosedPointIso_inv :
+    (stalkClosedPointIso R).inv = StructureSheaf.toStalk R _ := by
+  ext x
+  exact StructureSheaf.localizationToStalk_of _ _ _
+
+lemma ΓSpecIso_hom_stalkClosedPointIso_inv :
+    (Scheme.ΓSpecIso R).hom ≫ (stalkClosedPointIso R).inv =
+      (Spec R).presheaf.germ ⊤ (closedPoint _) trivial := by
+  rw [stalkClosedPointIso_inv, ← Iso.eq_inv_comp]
+  rfl
+
+@[reassoc (attr := simp)]
+lemma germ_stalkClosedPointIso_hom :
+    (Spec R).presheaf.germ ⊤ (closedPoint _) trivial ≫ (stalkClosedPointIso R).hom =
+      (Scheme.ΓSpecIso R).hom := by
+  rw [← ΓSpecIso_hom_stalkClosedPointIso_inv, Category.assoc, Iso.inv_hom_id, Category.comp_id]
+
+lemma Spec_stalkClosedPointIso :
```

Comment:
```suggestion
lemma Spec_map_stalkClosedPointIso :
```

Reply:
Or maybe `Spec.map_stalkClosedPointIso`?

Reply:
There are more occurrences below. I think we should decide whether `Spec_map_`, `Spec.map_` or `Spec_` is the convention.

Reply:
Probably `Spec_map_` I guess?